contracts/chain-adapters/utilities/HubPoolStore.sol

```diff
@@ -0,0 +1,128 @@
+     * be relayed to the UniversalStorageProof_SpokePool via storage proofs.
+     * Each call gets a unique nonce to prevent replay attacks on the L2 spoke pool.
+     * @dev Only callable by the HubPool contract.
+     * @param target Address of the contract on the destination that will receive the message.
+     * @param data Data to send to target.
+     */
+    function storeRelayAdminFunctionCalldata(address target, bytes calldata data) external onlyHubPool {
+        // Because we do not have the chain ID where the target is deployed, we can only associate this message
+        // with the target address. Therefore we are assuming that target spoke pool addresses are unique across
+        // chains. This technically doesn't prevent replay attacks where two spokes on different chains have
+        // the same address, but one way to prevent that is to deploy this store once per different L2 chain
+        // The L2's storage proof verification would only validate storage slots on this specific contract deployment.
+        _storeData(target, dataUuid++, data);
+    }
+
+    /**
+     * @notice To be called by HubPool to store calldata for the relayRootBundle function that will be relayed
+     * to the UniversalStorageProof_SpokPool via storage proofs. This function is a special
+     * case that is expected to be called once per L2 spoke pool using this adapter with the same calldata. Therefore,
+     * it is a gas optimization to call this function instead of storeRelayAdminFunctionCalldata because this function
+     * will only write a storage slot once per executed root bundle.
+     * @dev Only callable by the HubPool contract.
+     * @param target Address of the contract on the destination that will receive the message.
+     * @param data Data to send to target.
+     */
+    function storeRelayRootsCalldata(address target, bytes calldata data) external onlyHubPool {
+        HubPoolInterface.RootBundle memory pendingRootBundle = IHubPool(hubPool).rootBundleProposal();
+        // We need to know whether the data contains an admin or a normal relayRootBundle call in order to set the
+        // target and the nonce correctly in the data hash. The target cannot be set to address(0) for an admin bundle
+        // because then the admin bundle could be replayed on any spoke pool, and we should use the global nonce
+        // like other admin functions. Therefore, we check whether there is a pending bundle that hasn't passed
+        // liveness yet, or there is no bundle proposed. In these two cases, it is impossible for the HubPool
+        // to attempt to relay a root bundle unless it is relaying an admin bundle. In the third case where
+        // the pending bundle has passed liveness, we check whether the data is the same as the pending bundle
+        // to determine whether it is an admin bundle or not, as a final check.
+        bool isAdminRootBundle = pendingRootBundle.challengePeriodEndTimestamp > block.timestamp ||
```

Comment:
This is a pretty complex `bool` but it should protect us from sending admin relayRootBundle() calls that (1) can be replay attacked on different `UniversalStorageProof_SpokePools` and (2) block non-admin relayRootBundle() calls with identical calldata

Reply:
Yeah, they can, but we're checking that the multisig made the call to the HubPool, this logic wouldn't care who called the multisig.

Reply:
also this means we can't really handle the case well where the hub pool owner executes a root bundle...seems unlikely though

Reply:
https://github.com/across-protocol/contracts/pull/916/commits/130bb2fcd610d8b096182b5e1c7ee910278afda4